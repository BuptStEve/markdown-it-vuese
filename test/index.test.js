import fs from 'fs-extra'
import Md from 'markdown-it'
import path from 'path'
import vuese from '../src/'

describe('markdown-it-vuese', () => {
    const md = Md().use(vuese)

    describe('should', () => {
        it('render full documentation', async () => {
            const output = await renderOutput('at-path', md)

            expect(output).toMatchSnapshot()
        })

        it('render by custom root', async () => {
            const md = Md().use(vuese, {
                root: process.cwd() + '/test',
                useParser: parserRes => parserRes.name,
            })
            const output = await renderOutput('custom-root', md)

            expect(output).toBe(`<p>MyComponent</p>\n<p>MyComponent</p>`)
        })

        it('render by useParser', async () => {
            const md = Md().use(vuese, {
                useParser: parserRes => parserRes.name,
            })
            const output = await renderOutput('relative-path', md)

            expect(output).toBe(`<p>MyComponent</p>`)
        })

        it('render by useRender', async () => {
            const md = Md().use(vuese, {
                useRender: (vueseRender) => {
                    let content = ''
                    const renderRes = vueseRender.render()

                    Object.keys(renderRes).forEach((key) => {
                        content += `### ${key}\n\n${renderRes[key]}\n`
                    })

                    return content
                },
            })
            const output = await renderOutput('relative-path', md)

            expect(output).toMatchSnapshot()
        })

        it('render by useAll', async () => {
            const md = Md().use(vuese, {
                useAll: ({ markdownRes }) => markdownRes.componentName,
            })
            const output = await renderOutput('relative-path', md)

            expect(output).toBe(`<p>MyComponent</p>`)
        })

        it('match by custom vueseRe', async () => {
            const md = Md().use(vuese, {
                vueseRe: /\+{3}my-rule\((.+)\)/im,
                useAll: ({ markdownRes }) => markdownRes.componentName,
            })
            const output = await renderOutput('custom-re', md)

            expect(output).toBe(`<p>MyComponent</p>`)
        })

        it('render parse error', async () => {
            const output = await renderOutput('error', md)

            expect(output).toBe(`<p>Unexpected token, expected &quot;,&quot; (12:0)</p>`)
        })

        it('render multiple times', async () => {
            const md = Md().use(vuese, {
                useAll: ({ markdownRes }) => markdownRes.componentName,
            })
            const output = await renderOutput('multiple-times', md)

            expect(output).toBe(`<p>MyComponent</p>\n<p>MyComponent</p>`)
        })
    })

    describe('should not', () => {
        it('match fence', async () => {
            const output = await renderOutput('not-match', md)

            expect(output).toBe(`<pre><code class="language-md">&lt;[vuese](not-match.vue)\n</code></pre>`)
        })

        it('match code block', async () => {
            const output = await renderOutput('code-block', md)

            expect(output).toBe(`<pre><code>&lt;[vuese](@/test/MyComponent.vue)\n</code></pre>`)
        })

        it('found', async () => {
            const output = await renderOutput('not-found', md)

            expect(output).toBe(`<p>Not found: ${process.cwd()}/not-found.vue</p>`)
        })
    })

    // helpers

    async function renderOutput (name, md) {
        const input = await readFile(name)
        return md.render(input).trim()
    }

    async function readFile (name) {
        const target = path.resolve(__dirname, `./fragments/${name}.md`)
        return fs.readFile(target, 'utf-8')
    }
})
