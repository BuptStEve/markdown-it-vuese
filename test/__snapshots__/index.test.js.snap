// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown-it-vuese should render by useRender 1`] = `
"<h3>props</h3>
<table>
<thead>
<tr>
<th>Name</th>
<th>Description</th>
<th>Type</th>
<th>Required</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td>name</td>
<td>The name of the form, up to 8 characters</td>
<td><code>String</code> / <code>Number</code></td>
<td><code>true</code></td>
<td>-</td>
</tr>
</tbody>
</table>
<h3>slots</h3>
<table>
<thead>
<tr>
<th>Name</th>
<th>Description</th>
<th>Default Slot Content</th>
</tr>
</thead>
<tbody>
<tr>
<td>header</td>
<td>Form header</td>
<td><code>&lt;th&gt;title&lt;/th&gt;</code></td>
</tr>
</tbody>
</table>
<h3>events</h3>
<table>
<thead>
<tr>
<th>Event Name</th>
<th>Description</th>
<th>Parameters</th>
</tr>
</thead>
<tbody>
<tr>
<td>onclear</td>
<td>Fire when the form is cleared</td>
<td>The argument is a boolean value representing xxx</td>
</tr>
</tbody>
</table>
<h3>methods</h3>
<table>
<thead>
<tr>
<th>Method</th>
<th>Description</th>
<th>Parameters</th>
</tr>
</thead>
<tbody>
<tr>
<td>clear</td>
<td>Used to manually clear the form</td>
<td>-</td>
</tr>
</tbody>
</table>"
`;

exports[`markdown-it-vuese should render full documentation 1`] = `
"<h1>MyComponent</h1>
<h2>Props</h2>
<!-- @vuese:MyComponent:props:start -->
<table>
<thead>
<tr>
<th>Name</th>
<th>Description</th>
<th>Type</th>
<th>Required</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td>name</td>
<td>The name of the form, up to 8 characters</td>
<td><code>String</code> / <code>Number</code></td>
<td><code>true</code></td>
<td>-</td>
</tr>
</tbody>
</table>
<!-- @vuese:MyComponent:props:end -->
<h2>Events</h2>
<!-- @vuese:MyComponent:events:start -->
<table>
<thead>
<tr>
<th>Event Name</th>
<th>Description</th>
<th>Parameters</th>
</tr>
</thead>
<tbody>
<tr>
<td>onclear</td>
<td>Fire when the form is cleared</td>
<td>The argument is a boolean value representing xxx</td>
</tr>
</tbody>
</table>
<!-- @vuese:MyComponent:events:end -->
<h2>Slots</h2>
<!-- @vuese:MyComponent:slots:start -->
<table>
<thead>
<tr>
<th>Name</th>
<th>Description</th>
<th>Default Slot Content</th>
</tr>
</thead>
<tbody>
<tr>
<td>header</td>
<td>Form header</td>
<td><code>&lt;th&gt;title&lt;/th&gt;</code></td>
</tr>
</tbody>
</table>
<!-- @vuese:MyComponent:slots:end -->
<h2>Methods</h2>
<!-- @vuese:MyComponent:methods:start -->
<table>
<thead>
<tr>
<th>Method</th>
<th>Description</th>
<th>Parameters</th>
</tr>
</thead>
<tbody>
<tr>
<td>clear</td>
<td>Used to manually clear the form</td>
<td>-</td>
</tr>
</tbody>
</table>
<!-- @vuese:MyComponent:methods:end -->"
`;
